src/internal/clusterstate/v2.6.0/pfsdb.go

```diff
@@ -0,0 +1,580 @@
+	query := `SELECT int_id FROM pfs.commits WHERE commit_id = $1;`
+	err = tx.GetContext(ctx, &commitIntId, query, oldCommitKey(commit))
+	return errors.Wrapf(err, "get int_id of commit %q", oldCommitKey(commit))
+}
+
+func getFileSetMd(ctx context.Context, tx *pachsql.Tx, c *pfs.Commit) (*fileset.Metadata, error) {
+	var mdData []byte
+	if err := tx.GetContext(ctx, &mdData, `SELECT metadata_pb FROM storage.filesets JOIN pfs.commit_totals ON id = fileset_id WHERE commit_id = $1`, oldCommitKey(c)); err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+	md := &fileset.Metadata{}
+	if err := proto.Unmarshal(mdData, md); err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+	return md, nil
+}
+
+func sameFileSets(ctx context.Context, tx *pachsql.Tx, c1 *pfs.Commit, c2 *pfs.Commit) (bool, error) {
+	md1, err := getFileSetMd(ctx, tx, c1)
+	if err != nil {
+		return false, err
+	}
+	md2, err := getFileSetMd(ctx, tx, c1)
+	if err != nil {
+		return false, err
+	}
+
+	sameByteSlice := func(a []byte, b []byte) bool {
+		return true
+	}
+	sameIndex := func(a *idxProto.Index, b *idxProto.Index) bool {
+		if a.File != nil && b.File != nil && len(a.File.DataRefs) == len(b.File.DataRefs) {
+			for i, dr := range a.File.DataRefs {
+				if res := bytes.Compare(dr.Hash, b.File.DataRefs[i].Hash); res != 0 {
+					return false
+				}
+			}
+			return true
+		}
+		if a.Range != nil && b.Range != nil && sameByteSlice(a.Range.ChunkRef.Hash, b.Range.ChunkRef.Hash) {
+			return true
+		}
+		return false
+	}
+	// TODO(acohen4): add another utility ref that builds a composite hash
+	if md1.GetPrimitive() != nil && md2.GetPrimitive() != nil {
+		if !sameIndex(md1.GetPrimitive().GetAdditive(), md2.GetPrimitive().GetAdditive()) || !sameIndex(md1.GetPrimitive().GetDeletive(), md2.GetPrimitive().GetDeletive()) {
+			return false, nil
+		}
+		return true, nil
+	} else if md1.GetComposite() != nil && md2.GetComposite() != nil {
+		comp1Layers := md1.GetComposite().Layers
+		comp2Layers := md2.GetComposite().Layers
+		if len(comp1Layers) != len(comp2Layers) {
+			return false, nil
+		}
+		for i, l := range comp1Layers {
+			if l != comp2Layers[i] {
+				return false, nil
+			}
+		}
+		return true, nil
+	} else {
+		return false, nil
+	}
+}
+
+func convertCommitInfoToV2_6_0(ci *common.CommitInfo_V2_5_0) *pfs.CommitInfo {
+	return &pfs.CommitInfo{
+		Commit:              ci.Commit,
+		Origin:              ci.Origin,
+		Description:         ci.Description,
+		ParentCommit:        ci.ParentCommit,
+		ChildCommits:        ci.ChildCommits,
+		Started:             ci.Started,
+		Finishing:           ci.Finishing,
+		Finished:            ci.Finished,
+		Error:               ci.Error,
+		SizeBytesUpperBound: ci.SizeBytesUpperBound,
+		Details:             ci.Details,
+	}
+}
+
+func migrateAliasCommits(ctx context.Context, tx *pachsql.Tx) error {
+	var oldCIs []*common.CommitInfo_V2_5_0
+	var err error
+	// first fill commit provenance to all the existing commit sets
```

Comment:
It would be a bit clearer to separate the filling in of the new commits and provenance table from the removal of the alias commits.